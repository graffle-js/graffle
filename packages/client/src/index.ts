// Main client exports
export * from './client.js'
export * from './handle.js'
export * from './send.js'
