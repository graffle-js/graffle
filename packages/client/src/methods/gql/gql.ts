import type { Grafaid } from '@graffle/graphql'
import type { Context } from '@graffle/core/$.js'
import type { Configuration } from '@graffle/core/fragments/configuration/$.js'
import type { Docpar } from '@graffle/document/$.js'
import type { TypedFullDocument } from '@graffle/graphql/typed-full-document/$.js'
import type { ParseGraphQLObject, ParseGraphQLString } from '#src/static/gql.js'
import type { GlobalRegistry } from '#src/types/GlobalRegistry/GlobalRegistry.js'

import type { DocumentNode } from '@0no-co/graphql.web'
import type { DocumentSender, UntypedSender } from './DocumentSender.js'

/**
 * Validates that documents requiring SDDM are only used with SDDM-enabled contexts.
 *
 * This is a compile-time check that prevents documents generated with SDDM
 * (Schema-Driven Data Map) from being executed by clients without SDDM support.
 * SDDM documents contain type metadata and custom scalar information that require
 * the full schema to be available for proper encoding/decoding.
 *
 * Note: Documents with RequiresSDDM=false can be used with any client.
 * Documents with RequiresSDDM=true require clients configured with schema.map.
 */
// dprint-ignore
type ValidateSDDMRequirement<$Document extends Grafaid.Document.Typed.TypedDocumentLike, $Context> =
  Grafaid.Document.Typed.RequiresSDDMOf<$Document> extends true
    ? Configuration.Schema.HasMap<$Context> extends true
      ? $Document
      : `Error: This document requires SDDM but your client configuration lacks it.`
    : $Document

/**
 * Extract schema and arguments map from context for document object inference.
 *
 * @see {@link Configuration.Schema.Info}
 */
export type GetSchemaInfo<$Context> = Configuration.Schema.Info<$Context>

/**
 * Extract GraphQLStringAPI type from context for LSP detection.
 *
 * This type provides the `scalar`, `persisted`, and `__name` properties that GraphQLSP LSP
 * looks for to identify gql functions and determine which schema to use in multi-schema mode.
 *
 * Returns `any` when no GlobalRegistry is available (dynamic clients).
 */
// dprint-ignore
type StringAPIFromContext<$Context> = GlobalRegistry.ForContext<$Context> extends never
  ? any
  : Docpar.String.GraphQLStringAPI<
      Docpar.String.schemaOfSetup<{
        introspection: GlobalRegistry.ForContext<$Context>['stringIntrospection']
        scalars: GlobalRegistry.ForContext<$Context>['schema']['scalarRegistry']['map']
      }>,
      { isMaskingDisabled: false }
    >

/**
 * Check if GlobalRegistry is configured for this context.
 */
// dprint-ignore
type HasGlobalRegistry<$Context> =
  GlobalRegistry.ForContext<$Context> extends never
    ? false
    : true

/**
 * Extract document object constraint from GlobalRegistry if available.
 */
// dprint-ignore
type DocumentObjectConstraint<$Context> =
  HasGlobalRegistry<$Context> extends true
    ? GlobalRegistry.ForContext<$Context> extends { selectionSets: { $Document: infer $Doc } }
      ? $Doc
      : never
    : never

// dprint-ignore
/**
 * Execute a GraphQL document using GraphQL syntax or document builder.
 *
 * Accepts:
 * - Typed GraphQL documents (from codegen or static builders)
 * - Inline document builder objects
 *
 * Returns a {@link DocumentSender} with operation methods.
 *
 * The builder provides operation methods (one per operation in the document) that can be called
 * directly, or you can use the `.$send()` method to execute an operation by name.
 *
 * @example
 * ```ts
 * // GraphQL document string (with type inference)
 * const builder = graffle.gql('{ pokemons { name } }')
 * const data = await builder.myQuery()
 * ```
 *
 * @example
 * ```ts
 * // Inline document object
 * const data = await graffle.gql({ query: { myQuery: { id: true } } }).$send()
 * ```
 *
 * @remarks
 * TODO: Currently relies on $Document's default $DefaultSelectionContext parameter.
 * This works for statically-defined scalars but may not support runtime-added scalars
 * via `.use({ scalars: ... })`. If dynamic scalar context becomes a requirement,
 * revisit with HKT approach:
 * - Create SelectionSets TypeFunction that accepts context parameter
 * - Use TypeFunction.Call to apply runtime context to SelectionSets
 * - Extract $Document from result with applied context
 */
// dprint-ignore
export interface GqlMethod<$Context extends Context.Context> {
  // Overload
  // dprint-ignore
  <const $doc extends Grafaid.Document.Typed.String | string | DocumentNode | TypedFullDocument.TypedFullDocument>(
    document: ValidateSDDMRequirement<$doc, $Context>
  ):
  string extends $doc                                   ?
    $doc extends Grafaid.Document.Typed.String            ? DocumentSender<$doc, $Context> :
                                                            UntypedSender<$Context> :
  $doc extends TypedFullDocument.TypedFullDocument      ? DocumentSender<$doc, $Context> :
  $doc extends string                                   ? HasGlobalRegistry<$Context> extends true
                                                            ? ParseGraphQLString<$Context, $doc> extends infer $Parsed
                                                              ? $Parsed extends { __typename: 'ParserError' }
                                                                ? $Parsed
                                                                : $Parsed extends TypedFullDocument.TypedFullDocument
                                                                  ? DocumentSender<$Parsed, $Context>
                                                                  : never
                                                              : never
                                                            : UntypedSender<$Context> :
  $doc extends Grafaid.Document.Typed.TypedDocumentLike ? DocumentSender<$doc, $Context> :
                                                          never

  // Overload: Inline document builder object (must be last as it's least specific)
  <$Document extends DocumentObjectConstraint<$Context>>(
    document: $Document
  ): ParseGraphQLObject<$Context, $Document> extends infer $Parsed
    ? $Parsed extends { __typename: 'ParserError' }
      ? $Parsed
      : $Parsed extends TypedFullDocument.TypedFullDocument
        ? DocumentSender<$Parsed, $Context>
        : never
    : never

  /**
   * LSP detection property - identifies the schema name for multi-schema support.
   *
   * This property is used by GraphQLSP LSP in multi-schema mode to determine which
   * schema to use for validation and autocomplete. When multiple schemas are configured
   * in tsconfig.json, the LSP extracts this value to look up the appropriate schema.
   *
   * @see https://github.com/0no-co/GraphQLSP/blob/main/packages/graphqlsp/src/ast/checks.ts#L100-L124
   */
  readonly __name: Configuration.Schema.Name<$Context>

  /**
   * LSP detection property - validates scalar/enum values against schema types.
   *
   * This property is required by GraphQLSP LSP to identify this as a gql function.
   * The LSP checks for the existence of both `scalar` and `persisted` properties.
   *
   * @see https://github.com/0no-co/GraphQLSP/blob/main/packages/graphqlsp/src/ast/checks.ts#L19-L32
   */
  scalar: StringAPIFromContext<$Context>['scalar']

  /**
   * LSP detection property - creates persisted document nodes.
   *
   * This property is required by GraphQLSP LSP to identify this as a gql function.
   * The LSP checks for the existence of both `scalar` and `persisted` properties.
   *
   * @see https://github.com/0no-co/GraphQLSP/blob/main/packages/graphqlsp/src/ast/checks.ts#L19-L32
   */
  persisted: StringAPIFromContext<$Context>['persisted']
}

export namespace GqlMethod {
  /**
   * Arguments accepted by the instance gql method.
   *
   * Can be either:
   * - TypedDocumentLike (TypedDocumentNode, TypedDocumentString, etc.)
   * - Document builder object
   */
  export type Arguments =
    | [document: Grafaid.Document.Typed.TypedDocumentLike]
    | [documentObject: object]

  export type NormalizedArguments =
    | { type: 'typedDocument'; document: Grafaid.Document.Typed.TypedDocumentLike }
    | { type: 'object'; document: object }

  export const normalizeArguments = (args: Arguments): NormalizedArguments => {
    const first = args[0]

    // TypedDocumentLike documents are strings or have specific properties
    const isTypedDocumentLike = typeof first === 'string'
      || 'definitions' in first // DocumentNode
      || '__meta__' in first // TypedDocumentNode

    if (isTypedDocumentLike) {
      return {
        type: 'typedDocument',
        document: first as Grafaid.Document.Typed.TypedDocumentLike,
      }
    }

    return {
      type: 'object',
      document: first,
    }
  }
}
