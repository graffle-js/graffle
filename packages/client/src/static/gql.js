import { Docpar } from '@graffle/document/$.js';
import { print } from '@0no-co/graphql.web';
/**
 * Namespace for static gql function types and utilities.
 *
 * Provides the canonical input type definition for static document building.
 * For instance gql (which accepts pre-built typed documents), see GqlMethod.Arguments.
 */
export var Gql;
(function (Gql) {
    Gql.normalizeArguments = (args) => {
        const [first, second] = args;
        // String GraphQL document
        if (typeof first === 'string') {
            return {
                type: 'graphql',
                document: first,
                options: undefined,
            };
        }
        // Document builder object
        return {
            type: 'object',
            document: first,
            options: second,
        };
    };
})(Gql || (Gql = {}));
export const defaults = {
    hoistArguments: true,
};
export const createGql = (config) => {
    return ((...args) => {
        const normalized = Gql.normalizeArguments(args);
        if (normalized.type === 'graphql') {
            // For string inputs, return as-is
            // Type inference is handled at compile time by the document parser
            return normalized.document;
        }
        // Normalize the document object into internal representation
        const documentNormalized = Docpar.Object.Select.Document.normalizeOrThrow(normalized.document);
        // Convert to GraphQL document
        const result = Docpar.Object.ToGraphQLDocument.toGraphQLDocument(documentNormalized, {
            ...defaults,
            ...normalized.options,
            sddm: config.sddm,
        });
        // Print and return as TypedFullDocument
        return print(result.document);
    });
};
/**
 * Schema-less unified document builder function.
 *
 * Build GraphQL documents using either GraphQL string syntax or TypeScript objects.
 * Works without needing a client instance or schema generation, making it perfect for
 * quick prototyping, schema-less workflows, or when you prefer inline GraphQL strings.
 *
 * @remarks
 * This function accepts two input formats:
 * - **String**: Standard GraphQL document syntax with type inference
 * - **Object**: TypeScript objects with type-safe field selection
 *
 * For simpler single-field or multi-field operations, consider using the
 * {@link query} and {@link mutation} builders directly.
 *
 * @example GraphQL string documents
 * ```ts
 * import { gql } from 'graffle'
 *
 * // Simple query
 * const doc1 = gql(`query { user { id name } }`)
 *
 * // With variables
 * const doc2 = gql(`
 *   query GetUser($id: ID!) {
 *     user(id: $id) {
 *       id
 *       name
 *       email
 *     }
 *   }
 * `)
 *
 * // Multiple operations
 * const doc3 = gql(`
 *   query GetUser { user { id name } }
 *   mutation CreateUser { createUser { id } }
 * `)
 * ```
 *
 * @example Document object syntax
 * ```ts
 * import { gql } from 'graffle'
 *
 * // Single query operation
 * const doc1 = gql({
 *   query: {
 *     user: {
 *       id: true,
 *       name: true
 *     }
 *   }
 * })
 *
 * // Multiple operations
 * const doc2 = gql({
 *   query: {
 *     getUser: {
 *       user: { id: true, name: true }
 *     }
 *   },
 *   mutation: {
 *     createUser: {
 *       createUser: { id: true }
 *     }
 *   }
 * })
 * ```
 *
 * @see {@link https://graffle.js.org/guides/documents/static | Static Document Guide}
 * @see {@link query} For simplified query operations
 * @see {@link mutation} For simplified mutation operations
 */
export const gql = createGql({
    sddm: undefined, // Schema-less mode - no SDDM needed
});
//# sourceMappingURL=gql.js.map