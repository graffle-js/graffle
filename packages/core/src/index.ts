// Main core exports
export * from './context/context.js'
export * from './context/ContextEmpty.js'
export * as Context from './context/__.js'
export * from './types/ContextFragment.js'
export * as GlobalRegistry from './types/GlobalRegistry/types.js'
export * as RequestResult from './types/RequestResult/__.js'
export * from './utils.js'
