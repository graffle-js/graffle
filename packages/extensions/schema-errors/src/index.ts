export * from './runtime.js'
export * as Gentime from './gentime.js'
