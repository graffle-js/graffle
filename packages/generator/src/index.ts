// Main generator exports
export * from './generator/generate.js'
export * from './generator/config/config.js'
