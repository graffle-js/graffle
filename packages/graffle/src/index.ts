// Main graffle meta package - batteries included
export * from '@graffle/client'
export { DocumentBuilder } from '@graffle/extension-document-builder'
export { TransportHttp } from '@graffle/extension-transport-http'
export { TransportMemory } from '@graffle/extension-transport-memory'
export { Throws } from '@graffle/extension-throws'
export { Introspection } from '@graffle/extension-introspection'
export { SchemaErrors } from '@graffle/extension-schema-errors'
export { Opentelemetry } from '@graffle/extension-opentelemetry'
export { Upload } from '@graffle/extension-upload'
