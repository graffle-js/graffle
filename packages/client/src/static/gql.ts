import { Graphql, Docpar, Static } from '@graffle/graphql'
import { Core } from '@graffle/core'

// Re-export core gql functionality from graphql package
export const { createGql, gql, Gql, defaults } = Static
export type GqlInterface = typeof Static.gql

//
//
//
//
// ==================================================================================================
//                                    Context-Aware Parse Helpers
// ==================================================================================================
//
//
//
//

/**
 * Type-level utility that parses GraphQL string documents and returns a typed document with all operations.
 *
 * Extracts schema information from the context's GlobalRegistry and parses the GraphQL string accordingly.
 * If no GlobalRegistry is available, uses schema-less mode.
 *
 * @typeParam $Context - The client context containing schema introspection via GlobalRegistry
 * @typeParam $Input - The GraphQL document string to parse
 *
 * @example
 * ```ts
 * type Doc = ParseGraphQLString<MyContext, 'query MyQuery { id }'>
 * // Result: Document<Operation<'MyQuery', { id: string }, {}>>
 * ```
 */
// dprint-ignore
export type ParseGraphQLString<
  $Context,
  $Input extends string,
> = GlobalRegistry.ForContext<$Context> extends never
  ? Docpar.Parse<$Input, Docpar.ParserContext<undefined>>
  : Docpar.Parse<$Input, Docpar.ParserContext<GlobalRegistry.ForContext<$Context>['schema']>>

/**
 * Type-level utility that parses a document builder object and returns the typed document.
 *
 * Extracts schema and arguments map from the context's GlobalRegistry and infers operation types.
 *
 * @typeParam $Context - The client context containing schema and configuration via GlobalRegistry
 * @typeParam $Document - The document builder object to parse
 *
 * @example
 * ```ts
 * type Doc = ParseGraphQLObject<MyContext, { query: { myQuery: { id: true } } }>
 * // Result: Document with operation metadata
 * ```
 */
// dprint-ignore
export type ParseGraphQLObject<
  $Context,
  $Document,
> = Docpar.Parse<
  $Document,
  Docpar.ParserContext<
    GlobalRegistry.ForContext<$Context>['schema'],
    GlobalRegistry.ForContext<$Context>['argumentsMap'],
    $Context extends { typeHookRequestResultDataTypes: infer $TypeHooks } ? $TypeHooks : never
  >
>
