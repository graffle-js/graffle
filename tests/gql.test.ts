import gql from 'graphql-tag'
import { request } from '../src'
import { setupTestServer } from './__helpers'

const ctx = setupTestServer()

describe('gql', () => {
  it('passthrough allowing benefits of tooling for gql template tag', async () => {
    const mock = ctx.res({ body: { data: { foo: 1 } } })
    await request(
      ctx.url,
      gql`query allUsers {
  users
}
`
    )
    expect(mock).toMatchSnapshot()
  })
})
