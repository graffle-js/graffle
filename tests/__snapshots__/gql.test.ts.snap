// Vitest Snapshot v1

exports[`gql > passthrough allowing benefits of tooling for gql template tag 1`] = `
{
  "requests": [
    {
      "body": {
        "operationName": "allUsers",
        "query": "query allUsers {
  users
}",
      },
      "headers": {
        "accept": "*/*",
        "accept-encoding": "gzip,deflate",
        "connection": "close",
        "content-length": "67",
        "content-type": "application/json",
        "host": "DYNAMIC",
        "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
      },
      "method": "POST",
    },
  ],
  "spec": {
    "body": {
      "data": {
        "foo": 1,
      },
    },
  },
}
`;
